comid/classid.go

```diff
@@ -5,245 +5,375 @@ package comid
-		}
-		v = tnv{Type: "uuid", Value: b}
-	case TaggedOID:
-		b, err = OID(t).MarshalJSON()
+const IntType = "int"
+
+type TaggedInt int
+
+func NewIntClassID(val any) (*ClassID, error) {
+	if val == nil {
+		zeroVal := TaggedInt(0)
+		return &ClassID{&zeroVal}, nil
+	}
+
+	var ret TaggedInt
+
+	switch t := val.(type) {
+	case string:
+		i, err := strconv.Atoi(t)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("bad int: %w", err)
 		}
-		v = tnv{Type: "oid", Value: b}
-	case TaggedImplID:
-		b, err = ImplID(t).MarshalJSON()
-		if err != nil {
-			return nil, err
+		ret = TaggedInt(i)
+	case []byte:
+		if len(t) != 8 {
+			return nil, fmt.Errorf("bad int: want 8 bytes, got %d bytes", len(t))
 		}
-		v = tnv{Type: "psa.impl-id", Value: b}
+		ret = TaggedInt(binary.BigEndian.Uint64(t))
+	case int:
+		ret = TaggedInt(t)
+	case *int:
+		ret = TaggedInt(*t)
+	case int64:
+		ret = TaggedInt(t)
+	case *int64:
+		ret = TaggedInt(*t)
+	case uint64:
+		ret = TaggedInt(t)
+	case *uint64:
+		ret = TaggedInt(*t)
 	default:
-		return nil, fmt.Errorf("unknown type %T for class-id", t)
+		return nil, fmt.Errorf("unexpected type for int: %T", t)
 	}
 
-	return json.Marshal(v)
+	if err := ret.Valid(); err != nil {
+		return nil, err
+	}
+
+	return &ClassID{&ret}, nil
 }
 
-// Type returns the type of the target ClassID, i.e., one of UUID, OID or PSA
-// Implementation ID
-func (o ClassID) Type() ClassIDType {
-	switch o.val.(type) {
-	case TaggedUUID:
-		return ClassIDTypeUUID
-	case TaggedImplID:
-		return ClassIDTypeImplID
-	case TaggedOID:
-		return ClassIDTypeOID
-	}
-	return ClassIDTypeUnknown
+func (o TaggedInt) String() string {
+	return fmt.Sprint(int(o))
 }
 
-// String returns a printable string of the ClassID value. UUIDs use the
-// canonical 8-4-4-4-12 format, PSA Implementation IDs are base64 encoded.
-// OIDs are output in dotted-decimal notation.
-func (o ClassID) String() string {
-	switch t := o.val.(type) {
-	case TaggedUUID:
-		return UUID(t).String()
-	case TaggedImplID:
-		b := [32]byte(t)
-		return base64.StdEncoding.EncodeToString(b[:])
-	case TaggedOID:
-		return OID(t).String()
-	default:
-		return ""
-	}
+func (o TaggedInt) Valid() error {
+	return nil
+}
+
+func (o TaggedInt) Type() string {
+	return "int"
+}
+
+func (o TaggedInt) Bytes() []byte {
+	var ret [8]byte
+	binary.BigEndian.PutUint64(ret[:], uint64(o))
+	return ret[:]
 }
 
-// Unset tests whether the target ClassID has been initialized
-func (o ClassID) Unset() bool {
-	return o.val == nil || o.Type() == ClassIDTypeUnknown
+type IClassIDFactory func(any) (*ClassID, error)
+
+var classIDValueRegister = map[string]IClassIDFactory{
+	OIDType:  NewOIDClassID,
+	UUIDType: NewUUIDClassID,
+	IntType:  NewIntClassID,
+
+	ImplIDType: NewImplIDClassID,
+}
+
+// RegisterClassIDType registers a new IClassIDValue implementation (created
+// by the provided IClassIDFactory) under the specified CBOR tag.
+func RegisterClassIDType(tag uint64, factory IClassIDFactory) error {
```

Comment:
```suggestion
func RegisterClassIDType(tag uint64, newClassID IClassIDFactory) error {
```

Reply:
This would make it more confusing. The arguement is the the factory function, not a new class id. Something like `neewClassIDFunction` would make more sense, but it is more verbose, and again, "factory" is the standard well-understood term for the pattern being implemented here.